File: main_project/tripod-new/src/cpp/producermanager/TripodProducerManagerI.cpp
#include "TripodProducerManagerI.h"
#include "TripodProducer.h"
#include <Monitor/client/service/ServiceMonitor.h>

void MyUtil::initialize() {
    MyUtil::ServiceI& service = MyUtil::ServiceI::instance();
  service.getAdapter()->add(&com::renren::tripod::TripodProducerManagerI::instance(), service.createIdentity("M", ""));
}

namespace com {
namespace renren {
namespace tripod {

DataMap TripodProducerManagerI::produce(const KeySeq& keys, const std::string& business, bool isMaster, const Ice::Current& current) {
  std::ostringstream os;
  os<<"Key : ";
  for(KeySeq::const_iterator it = keys.begin(); it != keys.end(); ++it) {
      os<<*it<<" ";
  }
  os<<" business :"<<business;
  os<<" isMaster :"<<isMaster;
  MyUtil::InvokeClient ic = MyUtil::InvokeClient::create(current, os.str(), MyUtil::InvokeClient::INVOKE_LEVEL_INFO);

  CallFilter cf(filterHolder_, business);
  if(!cf.filter()) {
      return DataMap();
  }
 
  TripodProducerPtr producer = factory_.get(business);
  if (producer == NULL) {
    throw TripodProducerNotSupportedException();
  }
  return producer->create(keys, isMaster);
}

KeySeq TripodProducerManagerI::produceKeys(const std::string& beginKey, int limit, const std::string& business, bool hotKey, const Ice::Current& current) {
  std::ostringstream os;
  os<<"beginKey : "<< beginKey << " limit : " << limit << " business : " << business << " hotKey : " << hotKey;
  MyUtil::InvokeClient ic = MyUtil::InvokeClient::create(current, os.str(), MyUtil::InvokeClient::INVOKE_LEVEL_INFO);
  
  CallFilter cf(filterHolder_, business);
  if(!cf.filter()) {
      return KeySeq();
  }
 
  TripodProducerPtr producer = factory_.get(business);
  if (producer == NULL) {
    throw TripodProducerNotSupportedException();
  }
  if(hotKey) {
      return producer->createHotKeys(beginKey, limit);
  }else {
      return producer->createKeys(beginKey, limit);
  }
}


}
}
}

File: main_project/tripod-new-branch/src/java/TripodProducerManager/src/main/java/com/renren/tripod/TripodProducerManagerI.java
package com.renren.tripod;

import java.util.Map;

import Ice.Current;

public class TripodProducerManagerI extends _ProducerManagerDisp {
	/**
	 * 
	 */
	private static org.apache.commons.logging.Log _logger = org.apache.commons.logging.LogFactory
			.getLog(TripodProducerManagerI.class);
	private static final long serialVersionUID = 1L;
	
	private static CallFilter _callFilter = new CallFilter();

  private TripodProducerFactory _factory = new TripodProducerFactory();

	@Override
	public Map<String, String> produce(String[] keys, String business,
			boolean isMaster, Current current)
			throws TripodProducerNotSupportedException {
		_logger.info("TripodProducerManagerI.produce key size" + keys.length
				+ " business " + business + " isMaster " + isMaster);

		if(_callFilter.filter(business) == false) {
			return null;
		}
		
		try {
			TripodProducer producer = _factory.get(business);
			if (null == producer) {
				throw new TripodProducerNotSupportedException();
			}

			return producer.create(keys, isMaster);			
		}finally {
			_callFilter.exitFilter(business);
		}
	}

	@Override
	public String[] produceKeys(String beginKey, int limit, String business,
			boolean hotKey, Current current)
			throws TripodProducerNotSupportedException {
		_logger.info("TripodProducerManagerI.produceKeys beginKey " + beginKey
				+ " limit " + limit + " business " + business + " hotKey "
				+ hotKey);

		if(_callFilter.filter(business) == false) {
			return null;
		}
		
		try {
			TripodProducer producer = _factory.get(business);
			if (null == producer) {
				throw new TripodProducerNotSupportedException();
			}
			if (hotKey) {
				return producer.createHotKeys(beginKey, limit);
			} else {
				return producer.createKeys(beginKey, limit);
			}		
		}finally {
			_callFilter.exitFilter(business);
		}
		
	}

}
